python/lsst/dax/apdb/apdb.py

```diff
@@ -405,124 +493,103 @@ def getDiaSourcesInRegion(self, region: Region, dt: datetime
-        but types or units are not changed.
+    def _getSourcesByIDs(self, table: sqlalchemy.schema.Table,
+                         object_ids: List[int],
+                         midPointTai_start: float
+                         ) -> pandas.DataFrame:
+        """Returns catalog of DiaSource or DiaForcedSource instances given set
+        of DiaObject IDs.
 
         Parameters
         ----------
+        table : `sqlalchemy.schema.Table`
+            Database table.
         object_ids :
             Collection of DiaObject IDs
-        dt : `datetime.datetime`
-            Time of the current visit
+        midPointTai_start : `float`
+            Earliest midPointTai to retrieve.
 
         Returns
         -------
-        catalog : `pandas.DataFrame` or `None`
-            Catalog contaning DiaForcedSource records. `None` is returned if
+        catalog : `pandas.DataFrame`
+            Catalog contaning DiaSource records. `None` is returned if
             ``read_sources_months`` configuration parameter is set to 0 or
             when ``object_ids`` is empty.
         """
-
-        if self.config.read_forced_sources_months == 0:
-            _LOG.info("Skip DiaForceSources fetching")
-            return None
-
-        if len(object_ids) <= 0:
-            _LOG.info("Skip DiaForceSources fetching - no Objects")
-            # this should create a catalog, but the list of columns may be empty
-            return None
-
-        table: sqlalchemy.schema.Table = self._schema.forcedSources
         sources: Optional[pandas.DataFrame] = None
-
-        with Timer('DiaForcedSource select', self.config.timer):
-            with _ansi_session(self._engine) as conn:
+        with _ansi_session(self._engine) as conn:
+            if len(object_ids) <= 0:
+                _LOG.debug("ID list is empty, just fetch empty result")
+                query = table.select().where(False)
+                sources = pandas.read_sql_query(query, conn)
+            else:
                 for ids in _split(sorted(object_ids), 1000):
-
-                    query = 'SELECT *  FROM "' + table.name + '" WHERE '
+                    query = f'SELECT *  FROM "{table.name}" WHERE '
 
                     # select by object id
                     ids_str = ",".join(str(id) for id in ids)
-                    query += '"diaObjectId" IN (' + ids_str + ') '
+                    query += f'"diaObjectId" IN ({ids_str})'
+                    query += f' AND "midPointTai" > {midPointTai_start}'
 
                     # execute select
                     df = pandas.read_sql_query(sql.text(query), conn)
                     if sources is None:
                         sources = df
                     else:
                         sources = sources.append(df)
-
-        if sources is not None:
-            _LOG.debug("found %s DiaForcedSources", len(sources))
+        assert sources is not None, "Catalog cannot be None"
```

Comment:
Well, I think one of those two functions would still have to deal with None. I think the logic in this method is pretty clear, and I can convince myself easily that `sources` is indeed never `None`. `assert` is here mostly for `mypy`, and just in case we start messing with the code and break it accidentally. I'm happy with this `assert` if it keeps me from over-complicating things.